### ISSUE TITLE
Bug in DPOptimizer: actual sample size is not equal to batch size

### ISSUE THREAD

**Issue by charlotte12l** [pytorch/opacus#420]
## To Reproduce
> :warning: **We cannot help you without you sharing reproducible code**. Do not ignore this part :)

1. add `print('actual batch:', len(per_sample_clip_factor)` in  `clip_and_accumulate`
2. add `print('add noise!')` in `add_noise`
3. add `print('scale!', 'expected_batch_size:', self.expected_batch_size , 'iter:', self.accumulated_iterations)` in `scale_grad`

## Expected behavior

`actual sample size` is not equal to `expected_batch_size`! However, in `scale_grad`, it is using the `expected_batch_size` for average. Shouldn't we use the `actual sample size` for averaging? 

```
actual sample size: 14
add noise!
scale! expected_batch_size: 16 iter: 1
actual sample size: 10
add noise!
scale! expected_batch_size: 16 iter: 1
actual sample size: 17
add noise!
scale! expected_batch_size: 16 iter: 1


```

Besides, things would get worse when we use `BatchMemoryManager`, because we'll often do **two** rounds of `clip_and_accumulate` then do one round of `add_noise` and `scale_grad`. The `expected_batch_size` for average is totally different from the sum of the sample size feeded into the `clip_and_accumulate`
```
    with BatchMemoryManager(
        data_loader=train_dataloader, 
        max_physical_batch_size=MAX_PHYSICAL_BATCH_SIZE, 
        optimizer=optimizer
    ) as memory_safe_data_loader:
        for step, batch in enumerate(tqdm(memory_safe_data_loader)):

```

## Environment

I reproduced this use your template colab: [link](https://colab.research.google.com/drive/1LTAC_X8OMQh-x3gdt5H0HI8l_5uaMEDm?usp=sharing)

**Comment by karthikprasad** [comment]
Hi! The difference in actual batch size is expected and is due to Poisson sampling in DP Dataloader. See https://github.com/pytorch/opacus/blob/main/opacus/data_loader.py#L79-L82 
> ... it switches ``batch_sampler`` to ``UniformWithReplacementSampler``, thus enabling
    Poisson sampling (i.e. each element in the dataset is selected to be in the
    next batch with a certain probability defined by ``sample_rate`` parameter).
    NB: this typically leads to a batches of variable size.

**Comment by charlotte12l** [comment]
@karthikprasad  Thanks for your quick reply! I know you are doing Poisson sampling, but I think in the original DPSDG paper, they are doing average using the actual sample size? Or the Poisson sampling method actually refers to another paper?

Also, I know that either using `actual sample size` or `batch size` would not influence the performance much, because that is just equal to changing the learning rate. But I still think it would make more sense to use `actual sample size` for averaging? Thanks!

**Comment by karthikprasad** [comment]
This is based on [Mironov, Ilya, Kunal Talwar, and Li Zhang. ["Rényi Differential Privacy of the Sampled Gaussian Mechanism." arXiv preprint arXiv:1908.10530 (2019).](https://arxiv.org/abs/1908.10530)

**Comment by ffuuugor** [comment]
@charlotte12l actually, if my reading is correct, Abadi et al. paper also uses expected batch size in their theoretical work and only resort to some approximations for efficiency reasons in the practical part. 

in *Algorithm 1* they have group size *L* and dataset size *N*. They then define sampling probability by *L/N*, where *L* is the equivalent of our `expected_batch_size`. And then, if you look at **Add noise** section where they take an average of the gradients, they do so by dividing the sum of gradients by *L*, irrespective of the actual size of the lot *Lt*.

The authors later explicitly mention that it's not exactly what they do in practice for efficiency reasons:

> In practice, for efficiency,
the construction of batches and lots is done by randomly permuting the examples and then partitioning them into groups
of the appropriate sizes. For ease of analysis, however, we assume that each lot is formed by independently picking each
example with probability q = L/N, where N is the size of
the input datas

I'll close the issue for now, as the behaviour demonstrated by opacus looks correct in this case. Feel free to reopen the issue if you believe there's something wrong, or we can continue the discussion on [PyTorch forums](https://discuss.pytorch.org/c/opacus/29) if you have more questions

**Comment by charlotte12l** [comment]
@ffuuugor Sure! Thank you so much! I also modified the issue title. Sorry for the misleading title lol
